notifications/store/sql/store_provider.py

```diff
@@ -0,0 +1,237 @@
+"""
+Concrete MySQL implementation of the data provider interface
+"""
+
+import copy
+import pylru
+
+from django.core.exceptions import ObjectDoesNotExist
+
+from notifications.store.store import BaseNotificationStoreProvider
+from notifications.exceptions import (
+    ItemNotFoundError
+)
+from notifications import const
+from notifications.store.sql.models import (
+    SQLNotificationMessage,
+    SQLNotificationType,
+    SQLNotificationUserMap
+)
+
+
+class SQLNotificationStoreProvider(BaseNotificationStoreProvider):
+    """
+    Concrete MySQL implementation of the abstract base class (interface)
+    """
+
+    def __init__(self, **kwargs):
+        """
+        Initializer
+        """
+
+        _msg_type_cache_size = kwargs.get('MAX_MSG_TYPE_CACHE_SIZE', 1024)
+        self._msg_type_cache = pylru.lrucache(_msg_type_cache_size)
+
+    def _get_notification_by_id(self, msg_id, options=None):
+        """
+        Helper method to get Notification Message by id
+        """
+
+        # pylint/pep8 seem to complain if defaults are set to empty dicts
+        options = options if options else {}
```

Comment:
It flags it because mutable default args are really unintuitive. If `options` was defaulted to be an empty dict in the method signature, it would hold onto that reference across multiple method calls. So if you ever mutated `options` inside the method, every successive call to that method would have the same mutated value for `options`. It makes sense when you think about how the definition process actually works, but it bites every Python programmer that I know of at least once.